File: algorithms/is_unique.rb
# CtCI 6th Edition Problem 1.1
# Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you
# cannot use additional data structures?

# O(n) solution
# The string length will be the same as the set length if it is all unique characters
def unique_1?(str)
  str.split('').to_set.length == str.length
end

# O(n) solution
def unique_2?(str)
  visited_chars = Set.new
  str.each_char do |char|
    return false if visited_chars.include? char
    visited_chars << char
  end
  true
end

# Without data structures
# I cheat by turning this into an array to sort, but that's only because I am lazy and want to use build-in methods.
# This could be optimized by doing the comparison while sorting
def unique_3?(str)
  str.chars.sort.each_cons(2) { |char1, char2| return false if char1 == char2 }
  true
end

File: algorithms/is_unique.py
#!/usr/bin/env python3
# CtCI 6th Edition Problem 1.1
# Implement an algorithm to determine if a string has all unique characters.

# O(n^2) solution
def is_unique(string):
    for char in string:
        if string.count(char) > 1:
            return False
    return True

print(is_unique("test"))
print(is_unique("true"))

# I also saw this solution on SO that I really loved how concise it was.
# https://stackoverflow.com/a/35855968/4179267

# O(n) solution
def is_unique(string):
    return len(set(string)) == len(string)

print(is_unique("test"))
print(is_unique("true"))
